bcache.8
.TH bcache 8
.SH NAME
bcache \- manage bcache filesystems/devices

.SH  SYNOPSIS
.B bcache
[\fIoptions\fR]
.B COMMAND
[\fIoptions\fR]

.SH COMMANDS
.TP
.BR format
.RS
Format one or a list of devices with bcache data structures. You need to do this before you create a volume.
.RE

.BR register
.RS
Register a list of devices.
.RE

.BR list-cachesets
.RS
List cachesets on this system. This just searches /sys/fs/bcache.
.RE

.BR query-devs
.RS
Gives info about the superblock for a list of devices.
.RE

.BR status
.RS
Finds the status of the most up to date superblock in each tier. Pass in a list of devices.
.RE

.BR stats
.RS
List bcache specific statistics. Use --list to find out exactly what these are. To list stats for a cacheset, just give the cacheset uuid. To target a cache within the cacheset, use the -c option. To get time interval stats, use the hour, day, etc, options.
.RE

.BR probe
.RS
Return UUID if device identified as bcache-formatted.
.PP
Only necessary until support for the bcache superblock is included
in blkid; in the meantime, provides just enough functionality for a udev script
to create the /dev/disk/by-uuid symlink.
.RE

.BR help
.RS
List the bcache commands
.RE

.SH OPTIONS
.SH Options for bcache
.TP
.BR \--help


.SH Options for format
.SH Global options:
.TP
.BR \-w,\ --block=
block size, in bytes (e.g. 4k)
.TP
.BR \-w,\ --btree_node_size=
btree node size in bytes - default 256k
.TP
.BR \--metadata_checksum_type=TYPE
.TP
.BR \--data_checksum_type=TYPE
where TYPE is one of none, crc32c (default), or crc64
.TP
.BR \--compression_type=TYPE
where TYPE is one of none (default), lz4 or gzip
.TP
.BR \--encrypted
Enable encryption; passphrase will be prompted for
.TP
.BR \--error_action=TYPE
where TYPE is one of continue, readonly (default) or panic
.TP
.BR \-l\ --label=LABEL
Create the filesystem with the specified label
.TP
.BR \--uuid=UUID
Create the filesystem with the specified UUID
.TP
.BR \--force
Force the filesystem to be created, even if the device already contains a
filesystem

.SH Options that apply to subsequent devices:
.TP
.BR \--fs_size=SIZE
Create the filesystem using SIZE bytes on the subsequent device
.TP
.BR \-b\ --bucket-size=SIZE
Specifies the bucket size - must be greater than the btree node size
.TP
.BR \-t,\ --tier=INDEX
Specifies the tier of subsequent devices, where INDEX is a small integer and a
smaller index indicates a faster tier - tier 0 being the fastest. Currently only
two tiers are supported.
.TP
.BR \--discard
Enable discards on subsequent devices

.SH Options for register

.SH Options for list-cachesets
.TP
.BR \-d,\ --dir=
Provide a directory other than /sys/fs/bcache

.SH Options for query-devs
.TP
.BR \-f,\ --force-csum
Enables bcache to going even if the superblock crc is invalid

.SH Options for status

.SH Options for stats
.TP
.BR \-a,\ --all
List all stats and their values for the given device.
.TP
.BR \-l,\ --list
List the names of the different stats for the given device.
.TP
.BR \-u,\ --set=
UUID for the cacheset
.TP
.BR \-d,\ --dev=
UUID for a device within the cacheset. Must pass in the cacheset UUID in order to search for a dev.
.TP
.BR \-c,\ --cache=
Cache number, starts from 0
.TP
.BR \--five-min-stats
stats accumulated in the last 5 minutes
.TP
.BR \--hour-stats
stats accumulated in the last hour
.TP
.BR \--day-stats
stats accumulated in the last day
.TP
.BR \--total-stats
stats accumulated in total


debian/changelog
bcache-tools (1.0.8-2~bpo8+1) jessie-backports; urgency=medium

  * Rebuild for jessie-backports.

 -- Mathieu Parent <sathieu@debian.org>  Mon, 21 Sep 2015 21:18:39 +0200

bcache-tools (1.0.8-2) unstable; urgency=medium

  * Only run update-initramfs if installed. Fix dracut. (Closes: #788442)

 -- David Mohr <david@mcbf.net>  Thu, 11 Jun 2015 10:23:48 -0600

bcache-tools (1.0.8-1) unstable; urgency=medium

  [ James Page ]
  * d/control: Add Vcs fields.

  [ David Mohr ]
  * Don't depend on initramfs-tools, instead recommend it (Closes: #775674)
  * New upstream release 1.0.8
  * Update changelog
  * Add patch to clean bcache-register
  * Update changelog
  * Adding dep3 headers to the 0001 patch
  * Update watch to use http://evilpiepirate.org/git/bcache-tools.git
  * Add patch for gcc-5 compatability.
    Thanks to James Cowgill (Closes: #777798)

 -- David Mohr <david@mcbf.net>  Tue, 26 May 2015 20:57:58 -0600

bcache-tools (1.0.7-1) unstable; urgency=medium

  [ David Mohr ]
  * Based on work by Gabriel de Perthuis <g2p.code+debian@gmail.com>
  * Initial release. (Closes: #708132)

  [ Robie Basak ]
  * Remove unnecessary file bcache-tools.postrm.
  * debian/copyright fixes.
  * Add shebang to bcache-tools.preinst.
  * Drop Vcs-* for now.
  * Add self to Uploaders.

 -- Robie Basak <robie@justgohome.co.uk>  Mon, 27 Oct 2014 13:32:08 +0000


bcache.c
/*
 * Authors: Kent Overstreet <kent.overstreet@gmail.com>
 *	    Gabriel de Perthuis <g2p.code@gmail.com>
 *	    Jacob Malevich <jam@datera.io>
 *
 * GPLv2
 */

#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <inttypes.h>
#include <limits.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "bcache.h"

const char * const cache_state[] = {
	"active",
	"ro",
	"failed",
	"spare",
	NULL
};

const char * const replacement_policies[] = {
	"lru",
	"fifo",
	"random",
	NULL
};

const char * const csum_types[] = {
	"none",
	"crc32c",
	"crc64",
	NULL
};

const char * const compression_types[] = {
	"none",
	"lz4",
	"gzip",
	NULL
};

const char * const error_actions[] = {
	"continue",
	"readonly",
	"panic",
	NULL
};

const char * const bdev_cache_mode[] = {
	"writethrough",
	"writeback",
	"writearound",
	"none",
	NULL
};

const char * const bdev_state[] = {
	"detached",
	"clean",
	"dirty",
	"inconsistent",
	NULL
};

static void usage(void)
{
	puts("bcache - tool for managing bcache volumes/filesystems\n"
	     "usage: bcache <command> [<args>]\n"
	     "\n"
	     "Commands for formatting, startup and shutdown\n"
	     "  format         Format a new filesystem\n"
	     "  unlock         Unlock an encrypted filesystem prior to running/mounting\n"
	     "  assemble       Assemble an existing multi device filesystem\n"
	     "  incremental    Incrementally assemble an existing multi device filesystem\n"
	     "  run            Start a partially assembled filesystem\n"
	     "  stop	       Stop a running filesystem\n"
	     "\n"
	     "Commands for managing a running filesystem\n"
	     "  fs_show        Show various information about a filesystem\n"
	     "  fs_set         Modify filesystem options\n"
	     "\n"
	     "Commands for managing a specific device in a filesystem\n"
	     "  device_show    Show information about a formatted device\n"
	     "  device_add     Add a device to an existing (running) filesystem\n"
	     "  device_remove  Remove a device from an existing (running) filesystem\n");
	exit(EXIT_SUCCESS);
}

int main(int argc, char *argv[])
{
	char *cmd;

	if (argc < 2) {
		printf("%s: missing command\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	cmd = argv[1];

	memmove(&argv[1], &argv[2], argc * sizeof(argv[0]));
	argc--;

	if (!strcmp(cmd, "format"))
		return cmd_format(argc, argv);
	if (!strcmp(cmd, "assemble"))
		return cmd_assemble(argc, argv);
	if (!strcmp(cmd, "incremental"))
		return cmd_incremental(argc, argv);
	if (!strcmp(cmd, "run"))
		return cmd_run(argc, argv);
	if (!strcmp(cmd, "stop"))
		return cmd_stop(argc, argv);

	if (!strcmp(cmd, "fs_show"))
		return cmd_fs_show(argc, argv);
	if (!strcmp(cmd, "fs_set"))
		return cmd_fs_set(argc, argv);

	if (!strcmp(cmd, "device_show"))
		return cmd_device_show(argc, argv);
	if (!strcmp(cmd, "device_add"))
		return cmd_device_add(argc, argv);
	if (!strcmp(cmd, "device_remove"))
		return cmd_device_remove(argc, argv);

	if (!strcmp(cmd, "unlock"))
		return cmd_unlock(argc, argv);

	usage();
	return 0;
}
